compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -303,6 +308,7 @@ pub struct Body<'tcx> {
     /// and used for debuginfo. Indexed by a `SourceScope`.
     pub source_scopes: IndexVec<SourceScope, SourceScopeData<'tcx>>,
 
+    /// Additional information carried by a MIR body when it is lowered from a coroutine.
```

Comment:
Brief mention of when it goes from `Some` to `None` would be good.

Reply:
So this never goes from `Some` to `None` for normal MIR bodies. The only reason it "goes from `Some` to `None`" when we're building the coroutine drop shim is because we're cloning the whole MIR body struct (presumably out out of convenience) rather than, for example, just copying the basic blocks and locals into a fresh body.

The idea is that this is just basically just "take what we can from the parent MIR body, and then just fix up what is wrong". Other body information is patched up at the same time, such as the `source` and `arg_count`, and I don't believe that it makes sense to leave a parallel comment on `arg_count`, for example.

I can leave a comment on the `take` call below though.

Reply:
That would be great info to put into the comment! E.g. "`Some`" in the normal case, `None` only when we're building the coroutine drop shim".

Reply:
> None only when we're building the coroutine drop shim

And every other MIR body, not just coroutine drop shims. Which is why I think it feels redundant to note this, since this already says "when it is lowered from a coroutine".

Reply:
The final version looks fine, thanks!